### ISSUE TITLE
Provide generic API to ensime for other Vim plugins to integrate.

### ISSUE THREAD

**Issue by kstep** [ensime/ensime-vim#64]
I wasn't sure where should I start this discussion, so I decided to open an issue here.

The problem with current ensime-vim is it's trying to everything. E.g. it tries to highlight errors, but Syntastic plugin already dues it very well, much more better, than ensime-vim will be able to do in near future. I started an experiment in my fork (setloclist branch in kstep/ensime-vim), and now I have Syntastic and ensime play really well together!

The same thing is with YCM, which makes autocompletion very well (though it can integrate with ensime, if only it set omnifunc instead of completefunc). The same thing with Unite (I can imagine imports suggestions and symbols search with Unite, should be really awesome!).

I hope you got my point. The problem with ensime-vim, as I see it now, is it tries to play alone, doing things others already do much more better, than ensime-vim itself.

The one way to go out of this pit, as I see it, to make ensime-vim to provide generic Vim API to ensime calls, and then let community create integration plugins with other general-purpose plugins, like Syntastic, Unite etc.

I personally ready to provide Syntastic integration for errors highlighting, which already work for me very well.

**Comment by yazgoo** [comment]
Good idea.
Could you please provide an example of vim API.
Keeping in mind that ensime is asynchronous, so we should at least allow to register callback for specifics events.

**Comment by todesking** [comment]
I strongly agreed :+1: :+1: 
"Do one thing well" plugin is better than huge monolithic ones.

Maybe we need like this:
```vim
function IntegrateToSyntastic(path, errors)
  " ...
endfunction

call ensime#register_note_event(function('IntegrateToSyntastic'))
```

**Comment by kstep** [comment]
@todesking yes, that's one thing to do it! I thought about something more general, though, like:

```vim
call ensime#on_signal("NewErrors", function("AppendErrors"))
call ensime#on_signal("ErrorsComplete", function("ShowErrors"))
call ensime#emit_signal("GetErrors")
```

With possible helper functions like `fun ensime#on_errors()` and `fun ensime#get_errors()` which wrap more general forms.

I think I can do it, it can be simple enough refactoring from my Syntastic integration implementation.

**Comment by Ir1sh** [comment]
I really like this idea and would just like to make one comment regarding it.  How does providing an api for other existing plugins to interact with effect the 'onboarding' process for new people to either vim or ensime vim?  having to install ensime and ensime-vim and then go find each separate plugin that is going to interface with ensime-vim might be a pain.

I think the suggestion in the op is definitely the way to go but thinking about if its possible to 'package' all the dependencies to make start-up and maintenance less burdonsome is also worthwhile.

**Comment by kstep** [comment]
If you want nice syntax checker in vim, you will have Syntastic already installed, it's just a must have for a good vimmer. Generally, maybe ensime-vim should provide some basic things, like error highlighting, maybe by filling it quickfix location list (with `setqflist()`), so basic functionality is always here, but trying to do something more complex, IMHO, will result in code bloat and more problems, than benefits.

**Comment by yazgoo** [comment]
Here is a proposal: 

https://github.com/yazgoo/ensime-vim-typecheck

(via https://github.com/ensime/ensime-vim/pull/92 )

**Comment by g-rutter** [comment]
So would @kstep, or another user be able to explain how ensime-vim in its current form can be integrated with YCM and Syntastic?

It certainly seems clear that ensime should be integrating with these existing plugins through the standard routes, rather than trying to do all this stuff itself.

**Comment by ches** [comment]
@g-rutter Integration with YCM and Syntastic already exists and should "just work" in most cases. Sometimes completion plugins do need a bit of configuration tweaking, there are some issues you can find via search regarding deoplete settings, for instance (and if there are uncontroversial canonical settings we should include by default, we're open to that). I use YCM myself, just make sure `:setlocal omnifunc=EnCompleteFunc` and it ought to work (ensime-vim sets this automatically, but it's possible that some other plugin like Eclim may interfere and you'll need to work around it). Syntastic really should just work. Ask in Gitter if you have trouble.

Plugins like these already have their own integration APIs (or leverage Vim's, like `omnifunc`). In such cases ensime-vim ought to conform to them so users have little or nothing to do for them to work. This issue, to me, is about providing coherent scripting APIs and/or hooks for *new* plugins (or simply users' personal function libraries) to *use and extend ensime-vim features*. So it's the other way around.

Just to give a little bit of my current view on the standing of this matter in general:

@yazgoo did implement the basics of an event system similar to what's discussed above a long while ago, but it's basically undocumented, not covered by any current tests, and quite possibly broken after major refactorings of the plugin have happened and are still ongoing. So this issue remains open for discussion and improvements.

Our Vimscript function API is pretty lousy right now, I don't think it has ever really been "designed" consciously as an API, more of just mechanically exposing Python code to Vim on an as-needed basis for features to work. It's something that is on my mind to improve in time, but some pretty fundamental changes are being made to the core plugin now such that designing and committing to external APIs is likely to be a long way off.

If people have real motivating use cases for something like the event system proposed above, which might indeed be a good way to expose functionality for extension with a much easier to maintain interface for compatibility, I think the best way to approach a proposal is to demonstrate a proof-of-concept implementation of an integration/extension (like @yazgoo's linked repo above, but ideally with a real use case 😄 ), *and documentation and unit tests* for the event system or similar. We have a test suite going again now. It may well be that we could then start factoring some of the core functionality into "plugins" that are distributed with the core ensime-vim but are integrated just like third-party code would be—great!

I'd like to think about ways to use existing Vim idioms as much as possible, like user autocommands, but they are limited so that might not go very far.
